fuse_core/include/fuse_core/callback_wrapper.h

```diff
@@ -127,13 +141,59 @@ class CallbackWrapper : public ros::CallbackInterface
 // Specialization to handle 'void' return types
 // Specifically, promise_.set_value(callback_()) does not work if callback_() returns void.
 template <>
-inline ros::CallbackInterface::CallResult CallbackWrapper<void>::call()
+inline void CallbackWrapper<void>::call()
 {
   callback_();
   promise_.set_value();
-  return Success;
 }
 
+
+class CallbackAdapter : public rclcpp::Waitable
+{
+public:
+
+  CallbackAdapter(std::shared_ptr<rclcpp::Context> context_ptr);
+
+  /**
+   * @brief tell the CallbackGroup how many guard conditions are ready in this waitable
+   */
+  size_t get_number_of_ready_guard_conditions();
+
+
+  /**
+   * @brief tell the CallbackGroup that this waitable is ready to execute anything
+   */
+  bool is_ready(rcl_wait_set_t * wait_set);
+
+
+  /**
+   * @brief add_to_wait_set is called by rclcpp during NodeWaitables::add_waitable() and CallbackGroup::add_waitable()
+    waitable_ptr = std::make_shared<CallbackWrapper>();
+    node->get_node_waitables_interface()->add_waitable(waitable_ptr, (rclcpp::CallbackGroup::SharedPtr) nullptr);
+   */
+  void add_to_wait_set(rcl_wait_set_t * wait_set);
+
+  std::shared_ptr< void > take_data();
+
+  // XXX check this against the threading model of the multi-threaded executor.
+  void execute(std::shared_ptr<void> & /*data*/);
+
+  void addCallback(const std::shared_ptr<CallbackWrapperBase> &callback);
+
+  void addCallback(std::shared_ptr<CallbackWrapperBase> && callback);
+
+  void removeAllCallbacks();
+
+
+private:
+  std::recursive_mutex reentrant_mutex_;  //!< mutex to allow this callback to be added to multiple callback groups simultaneously
```

Comment:
Does "multiple callback groups simultaneously" mean adding the same Waitable instance to multiple executors? That sounds undesirable. If two executors are trying to execute the same thing at the same time then one is going to be blocked, when it could have been executing something else.

Can this be removed?

Reply:
the comment is mistaken, (oops)
reentrant_mutex is called  to protect `rcl_wait_set_add_guard_condition`, which inserts a pointer into a list of pointers and is not thread-safe.  This allows multiple threads to insert callbacks into one queue without trashing RCL's internal data structures

queue_mutex is the one you're worried about, it allows multiple executors to pull jobs from the queue.
If someone managed to the wrapper in multiple executors without the mutex, it would be a nightmare to debug.

The lock is released before the callback is run. An executor will only be blocked by the time it takes for another executor to pull a pointer from a queue.
https://github.com/locusrobotics/fuse/blob/ee39c45cb7a704e0c646daf53993cfb5085f0d7c/fuse_core/src/callback_wrapper.cpp#L102-L112
Executors won't try to execute the same code twice, and they won't wait for a callback to finish either.

Reply:
Updated the comment https://github.com/locusrobotics/fuse/pull/284/commits/fb5254343a1098ead1608a3d65580615e611f1d6